grpc-client-server/prober_server/main.go

```diff
@@ -9,37 +9,70 @@ import (
 	"net/http"
 	"time"
 
-	pb "github.com/CodeYourFuture/immersive-go-course/grpc-client-server/prober"
+	pb "github.com/Berkeli/immersive-go/grpc-client-server/prober"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
 	"google.golang.org/grpc"
 )
 
 var (
-	port = flag.Int("port", 50051, "The server port")
+	port         = flag.Int("port", 50051, "The server port")
+	LatencyGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "latency_gauge",
+		Help: "The latency of the requests to the endpoint",
+	}, []string{"endpoint"})
 )
 
 // server is used to implement prober.ProberServer.
-type server struct {
+type Server struct {
 	pb.UnimplementedProberServer
 }
 
-func (s *server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
+func (s *Server) DoProbes(ctx context.Context, in *pb.ProbeRequest) (*pb.ProbeReply, error) {
 	// TODO: support a number of repetitions and return average latency
-	start := time.Now()
-	_, _ = http.Get(in.GetEndpoint())	// TODO: add error handling here and check the response code
-	elapsed := time.Since(start)
-	elapsedMsecs := float32(elapsed / time.Millisecond)
+	totalMsecs := 0
+	failed := 0
+	for i := 0; i < int(in.GetNumberOfRequests()); i++ {
+		start := time.Now()
+		resp, err := http.Get(in.GetEndpoint())
+		if err != nil {
+			log.Printf("could not probe: %v", err)
+			failed++
+			continue
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != http.StatusOK {
+			log.Printf("Received status %d during probe", resp.StatusCode)
+			failed++
+			continue
+		}
+		elapsed := time.Since(start)
```

Comment:
There are two metrics we typically measure when probing like this: "Time to first byte" and "Time to last byte" - which are you measuring here? What are the benefits of both?

Reply:
I think my measurement is "time to last byte" as I'm only checking time after I have successfully received response. 

TTFB would be useful in measuring server responsiveness as that indicates the time required to establish a connection and start receiving data. TTLB can be misleading (for our usecase) as it would depend on the size of the webpage.  

For our usecase, TTFB is probably better since we are only checking response times and not the actual load times.

Reply:
How can you test whether you're measuring time to first byte or time to last byte? Why not try to write a test which proves which one you're doing :)

In general I'd say time to first byte is useful for discovering whether servers are _up_, and maybe some geographic testing, and time to _last_ byte is generally useful for seeing how a server is doing, and measuring user-facing performance. As a rule, unless your goal is just "is the server up", I'd lean towards time to _last_ byte.

Reply:
So after doing some research, I think I'm actually doing a TTFB measure in these cases or something in between ttfb and ttlb? 

If my understanding is correct - then TTFB happens slightly before we have all response headers and can start reading from the body. I have done some testing with httptrace and httpstat which uses httptrace under the hood.

not 100% sure if my implementation of TTLB is correct - waiting to read everything from the body.  And I guess the measurement should be taken before we close the connection?